### ISSUE TITLE
Bundle does not build gems in parallel

### ISSUE THREAD

**Issue by J0WI** [rubygems/rubygems#5276]
### Describe the problem as clearly as you can

Gems are always build with just one core even if the jobs option in bundler is set. Building them in parallel would be _a lot_ faster (basically multiplied by number of available CPU cores).

### Did you try upgrading rubygems & bundler?

yes

### Post steps to reproduce the problem

Run `bundle add`, `bundle install` or `bundle update` for any gem that requires a build step.

### Which command did you run?

e.g. `bundle install sassc`

### What were you expecting to happen?

From the docs:

```sh
* `--jobs=[<number>]`, `-j[<number>]`:
  The maximum number of parallel download and install jobs. The default is the
  number of available processors.
```

I'd expect that the gem is compiled using all available cores.

### What actually happened?

Only one core is used to build the gem. Even `bundle install -j 4 sassc` is still using just one core. For some gems or projects this process wastes _a lot_ of time.

The issue seems to in the make command of RubyGems:
https://github.com/rubygems/rubygems/blob/a9f9341e08478426ca3aa7829b5be4534cee06c1/lib/rubygems/ext/builder.rb#L38-L42

Manually injecting the `jobs` flag here fixes the problem for me. `build_args` are only applied to the `build function` but there is no equivalent for `make`:
https://github.com/rubygems/rubygems/blob/a9f9341e08478426ca3aa7829b5be4534cee06c1/lib/rubygems/ext/builder.rb#L161-L162

So AFAIK there is currently no way for Bundler to pass the parallel option to make.

I could only submit a patch that copies the behavior of Bundler in RubyGems:
https://github.com/rubygems/rubygems/blob/a9f9341e08478426ca3aa7829b5be4534cee06c1/bundler/lib/bundler/settings.rb#L211-L216

Ideally Bundler itself would pass the option to RubyGems.

### If not included with the output of your command, run `bundle env` and paste the output below

_not applicable_

**Comment by simi** [comment]
@J0WI this works as expected (at least to me). `--jobs` flag passed to `bundle` configures amount of threads working at the same time at installing individual gems. But there is currently no setup to pass `-j` to `make` command used in gems with `Makefile` extensions (like `sassc` gem in your example).

`bundle install -j 4 sassc` fails on my side, since `bundle install` doesn't expect gem name.

:information_source: It will make sense to enhance `gem install` command with new argument to make extension compilation done in parallel when possible. Later we can incorporate it to bundler as well to be able to configure this flag per gem using `bundle config command`.

**Comment by elthariel** [comment]
One way to achieve this is to use the `MAKE` env var:

`MAKE="make -j4" bundle install`

**Comment by simi** [comment]
@elthariel that seems related to this line:

https://github.com/rubygems/rubygems/blob/62461d6ab69a12806bf3ee84a454747a7cd3f1e1/lib/rubygems/ext/builder.rb#L27

The problem with "parallel gem install" combined with "parallel make" is (in theory) it can install 4 gem extensions at the same time using `make -j4` which will "utilise" 16 "cores" when only 4 were originally requested.

```ruby
MAKE="make -j4" bundle install -j 4
```

**Comment by elthariel** [comment]
@simi You are totally right. In my case, I don't really care saturating the running machine beyond what was request, I'm caring about NOT saturating it :)

But one should be aware of this behavior when deciding to use my tiny hack, so I thank you for that :)

**Comment by 64kramsystem** [comment]
> The MAKEFLAGS variable might also be of use, but I haven't tried it

As far as I've read, it's not an environment variable that the user can set, but it's used internally by make. The portability should also be verified - GNU Make and Nmake do support it, but I don't know about the rest :)

**Comment by 64kramsystem** [comment]
> @elthariel that seems related to this line:
> 
> https://github.com/rubygems/rubygems/blob/62461d6ab69a12806bf3ee84a454747a7cd3f1e1/lib/rubygems/ext/builder.rb#L27
> 
> The problem with "parallel gem install" combined with "parallel make" is (in theory) it can install 4 gem extensions at the same time using `make -j4` which will "utilise" 16 "cores" when only 4 were originally requested.
> 
> ```ruby
> MAKE="make -j4" bundle install -j 4
> ```

I have the suspicion that there is some complexity in this case.

Based on my understanding, this should not be a CPU problem due to the GIL. Bundler jobs are I/O constrainted (downloading, unpacking/copying etc.) so parallelism has a chance to occur, but extension compilation is (mostly) CPU bound, so there shouldn't be, at a point in time, more than one gem compiilng.

On the other hand, the compiler instances should still hang around in the (virtual) memory, which may be in theory flooded, so the exponentiation problem is there, in one form or the other :grimacing: On a 32 threads system, 1024 (32^2) compilation threads are a fun/scary prospect :clown_face:

**Comment by hsbt** [comment]
We added "-j" flags for C extension builder at https://github.com/ruby/rubygems/pull/9131. You can use it at RubyGems 4.0.
